### PR TITLE
fix(dataset): harden inputs.json and raw-payload loader input validation

### PR DESCRIPTION
**Pull request by ajcasagrande**
## Summary

Hardens input validation in two dataset loaders so malformed user inputs fail with actionable errors instead of bare tracebacks or silent misclassification. Carved from the `ajc/agentx-on-main` adversarial-hardening work; includes the two new adversarial test suites that pin the fixed behavior.

### 1. `InputsJsonPayloadLoader.load_dataset` — missing required keys (`src/aiperf/dataset/loader/inputs_json.py`)

An `inputs.json` entry missing `session_id` or `payloads` crashed with a bare, context-free exception:

```
KeyError: 'session_id'
```

Now raises a path- and index-rich error:

```
ValueError: <path>: data[0] missing required key 'session_id'
```

### 2. `RawPayloadDatasetLoader.can_load` — non-dict probe data (`src/aiperf/dataset/loader/raw_payload.py`)

Type-dispatch auto-detection feeds arbitrary first-record shapes into `can_load`. A non-dict record (list, string, scalar) crashed the detection chain mid-walk:

```
AttributeError: 'str' object has no attribute 'get'
```

Now guarded with `isinstance(data, dict)` so detection falls through cleanly (`can_load` returns `False`).

## Deliberate behavior change: OSError now propagates from `_dir_has_raw_payload_jsonl`

Previously `_dir_has_raw_payload_jsonl` caught `(orjson.JSONDecodeError, OSError)` and skipped to the next file. That meant an unreadable `.jsonl` file (e.g. permission-denied) was silently swallowed and the directory was **misclassified as not-raw-payload**, sending the run down the wrong loader path with no signal.

The except clause is narrowed to `(orjson.JSONDecodeError, ValueError)`: only malformed JSON is skipped; `OSError`/`PermissionError` from an unreadable file now **propagates out of `can_load`** so misconfigured inputs fail loudly at detection time. This is an intentional loud-failure change, pinned by `test_can_load_directory_with_unreadable_jsonl_raises_post_fix`.

## Tests

- `tests/unit/dataset/loader/test_inputs_json_adversarial.py` (new, 142 lines) — `can_load` boundary shapes, Pydantic `min_length` enforcement, session_id passthrough, one-turn-per-payload, duplicate/missing-key error pins.
- `tests/unit/dataset/loader/test_raw_payload_adversarial.py` (new, 218 lines) — non-dict probe data, directory-peek discrimination (empty/null/array first lines, wrong extensions, malformed JSONL), JSONL line edge cases (blank lines, trailing newline, verbatim `raw_payload`), unreadable-file loud-failure pin.

### Revert-verification (pin discipline)

Reverse-applied the two src fixes onto `main` and ran the adversarial suites — **5 tests failed, reproducing the exact pre-fix crashes**:

```
FAILED test_raw_payload_adversarial.py::TestCanLoadDataShape::test_can_load_non_dict_data_returns_false[bad_data0]
FAILED test_raw_payload_adversarial.py::TestCanLoadDataShape::test_can_load_non_dict_data_returns_false[string]
FAILED test_raw_payload_adversarial.py::TestCanLoadDataShape::test_can_load_non_dict_data_returns_false[123]
       -> AttributeError: 'str' object has no attribute 'get'  (raw_payload.py:54)
FAILED test_inputs_json_adversarial.py::TestWave2FixForwardCompatibility::test_load_dataset_missing_required_key_raises_value_error_post_fix
       -> bare KeyError: 'session_id' instead of ValueError
FAILED test_raw_payload_adversarial.py::TestWave2BugCandidates::test_can_load_directory_with_unreadable_jsonl_raises_post_fix
       -> OSError swallowed; can_load silently returned False
```

Re-applied the fixes: all green.

### Evidence

- `uv run pytest -n auto tests/unit/dataset/loader/test_inputs_json_adversarial.py tests/unit/dataset/loader/test_raw_payload_adversarial.py tests/unit/dataset/loader/test_inputs_json_payload.py tests/unit/dataset/loader/test_raw_payload.py` — **47 passed**
- `pytest --collect-only tests/unit/dataset/loader/` — clean (931 collected)
- `ruff format` + `ruff check` on all four touched files — clean
- `make check-ruff-baselined` — OK (118 total, 118 baselined, **0 new**)
- `make check-ergonomics` — OK (58 total, 58 baselined, **0 new**)

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **Bug Fixes**
  * Added upfront per-entry validation for Inputs JSON loading, including clear `ValueError` messages when required fields are missing or entries aren’t objects.
  * Strengthened raw payload detection by rejecting non-dictionary inputs earlier and ensuring unreadable/malformed directory entries are handled correctly (permission/read errors fail loudly).
  * Improved conversation conversion behavior with stricter validation and correct preservation of `raw_payload`.
* **Tests**
  * Added extensive adversarial unit tests for both loaders, including Wave 2 forward-compatibility and JSON/JSONL boundary cases.

### CODE REVIEW

File: src/aiperf/dataset/loader/inputs_json.py [hunk 1]
```diff
@@ -72,6 +72,12 @@ def load_dataset(self) -> dict[str, list[InputsJsonSession]]:
 
         result: dict[str, list[InputsJsonSession]] = {}
         for idx, entry in enumerate(data_list):
+            if "session_id" not in entry:
+                raise ValueError(
+                    f"{path}: data[{idx}] missing required key 'session_id'"
+                )
+            if "payloads" not in entry:
+                raise ValueError(f"{path}: data[{idx}] missing required key 'payloads'")
```
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply src/aiperf/dataset/loader/inputs_json.py:84]: Fixed: added an isinstance(entry, dict) guard that raises ValueError with the index and offending type before the key-presence checks.

File: tests/unit/dataset/loader/test_inputs_json_adversarial.py [hunk 2]
```diff
@@ -0,0 +1,142 @@
+# SPDX-FileCopyrightText: Copyright (c) 2025-2026 NVIDIA CORPORATION & AFFILIATES. All rights reserved.
+# SPDX-License-Identifier: Apache-2.0
+"""Adversarial coverage for InputsJsonPayloadLoader.
+
+Pins current behavior for edge cases in `can_load` and `load_dataset`,
+and marks two xfail-strict tests that will flip to pass when the known
+bugs (duplicate session_id overwrite, bare KeyError on missing keys) are
+fixed in Wave 2.
+"""
```
[the remark this answers is not in the document]
Comment (ajcasagrande) [reply tests/unit/dataset/loader/test_inputs_json_adversarial.py:9]: Fixed: updated the module docstring to describe the actual coverage (direct pytest.raises assertions for the Wave 2 fixes) and dropped the stale xfail-strict mention.

File: tests/unit/dataset/loader/test_inputs_json_adversarial.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+                {"session_id": "dup", "payloads": [{"model": "m2"}]},
+            ]
+        }
+        path = tmp_path / "inputs.json"
+        path.write_bytes(orjson.dumps(data))
+        loader = _make_loader(path)
+        with pytest.raises(ValueError, match="duplicate"):
+            loader.load_dataset()
+
+    def test_load_dataset_missing_required_key_raises_value_error_post_fix(
```
Comment (lkomali) [inline tests/unit/dataset/loader/test_inputs_json_adversarial.py:140]: This only drops `session_id`, so the new `if "payloads" not in entry: raise` branch in `load_dataset` is never exercised. Since the point of this PR is pinning the fixed behavior, that branch is worth covering too. Simplest is to parametrize over the dropped key:

```python
@pytest.mark.parametrize("missing_key", ["session_id", "payloads"])
def test_load_dataset_missing_required_key_raises_value_error_post_fix(self, tmp_path, missing_key):
    entry = {"session_id": "s", "payloads": [{"x": 1}]}
    del entry[missing_key]
    path = tmp_path / "inputs.json"
    path.write_bytes(orjson.dumps({"data": [entry]}))
    with pytest.raises(ValueError, match=missing_key):
        _make_loader(path).load_dataset()
```

File: src/aiperf/dataset/loader/raw_payload.py [hunk 4]
```diff
@@ -156,7 +161,10 @@ def _dir_has_raw_payload_jsonl(directory: Path) -> bool:
                     return isinstance(record, dict) and isinstance(
                         record.get("messages"), list
                     )
-        except (orjson.JSONDecodeError, OSError):
+        except (orjson.JSONDecodeError, ValueError):
```
Comment (lkomali) [inline src/aiperf/dataset/loader/raw_payload.py]: Minor: `orjson.JSONDecodeError` is already a subclass of `ValueError`, so listing both is redundant, `except ValueError` alone catches malformed JSON. Not important, keep it if you prefer the explicitness.

**Review by lkomali — APPROVED** [review]
Clean, well-scoped hardening, and the adversarial suites with revert-verification are solid. One test worth adding before merge, plus a small nit.
